datafusion/src/optimizer/simplify_expressions.rs

```diff
@@ -0,0 +1,506 @@
+        Expr::BinaryExpr { .. } => Some(expr),
+        _ => None,
+    }
+}
+
+fn operator_is_boolean(op: Operator) -> bool {
+    op == Operator::And || op == Operator::Or
+}
+
+fn is_one(s: &Expr) -> bool {
+    match s {
+        Expr::Literal(ScalarValue::Int8(Some(1)))
+        | Expr::Literal(ScalarValue::Int16(Some(1)))
+        | Expr::Literal(ScalarValue::Int32(Some(1)))
+        | Expr::Literal(ScalarValue::Int64(Some(1)))
+        | Expr::Literal(ScalarValue::UInt8(Some(1)))
+        | Expr::Literal(ScalarValue::UInt16(Some(1)))
+        | Expr::Literal(ScalarValue::UInt32(Some(1)))
+        | Expr::Literal(ScalarValue::UInt64(Some(1))) => true,
+        Expr::Literal(ScalarValue::Float32(Some(v))) if *v == 1. => true,
+        Expr::Literal(ScalarValue::Float64(Some(v))) if *v == 1. => true,
+        _ => false,
+    }
+}
+
+fn is_true(expr: &Expr) -> bool {
+    match expr {
+        Expr::Literal(ScalarValue::Boolean(Some(v))) => *v,
+        _ => false,
+    }
+}
+
+fn is_false(expr: &Expr) -> bool {
+    match expr {
+        Expr::Literal(ScalarValue::Boolean(Some(v))) => *v == false,
+        _ => false,
+    }
+}
+
+fn simplify(expr: &Expr) -> Expr {
+    match expr {
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Or,
+            right,
+        } if is_true(left) || is_true(right) => lit(true),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Or,
+            right,
+        } if is_false(left) => simplify(right),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Or,
+            right,
+        } if is_false(right) => simplify(left),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Or,
+            right,
+        } if left == right => simplify(left),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::And,
+            right,
+        } if is_false(left) || is_false(right) => lit(false),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::And,
+            right,
+        } if is_true(right) => simplify(left),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::And,
+            right,
+        } if is_true(left) => simplify(right),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::And,
+            right,
+        } if left == right => simplify(right),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Multiply,
+            right,
+        } if is_one(left) => simplify(right),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Multiply,
+            right,
+        } if is_one(right) => simplify(left),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Divide,
+            right,
+        } if is_one(right) => simplify(left),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Divide,
+            right,
+        } if left == right => lit(1),
```

Comment:
This isn't correct if both sides can be null.

Reply:
You're right, should be fixed now.

Reply:
I merged it, but see this part might still be not 100% correct.

if left and right are a column, the value can still be null (for some evaluations).
Also the division by 0 (e.g. 0/0 or x/x where x is 0) is tricky.
I think there is no easy way to do it safely.